video/out/opengl/hwdec_cuda.c

```diff
@@ -300,105 +288,193 @@ static void cuda_uninit(struct ra_hwdec *hw)
-            .handle.fd = ebuf->fd,
+        .type = CU_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_FD,
+        .handle.fd = wait_fd,
 #endif
-            .size = mem_info.mem_size,
-            .flags = 0,
-        };
-        ret = CHECK_CU(cu->cuImportExternalMemory(&ebuf->mem, &ext_desc));
-        if (ret < 0)
-            goto error;
+    };
+    ret = CHECK_CU(cu->cuImportExternalSemaphore(&evk->ws, &w_desc));
+    if (ret < 0)
+        goto error;
+    wait_fd = -1;
 
-        CUDA_EXTERNAL_MEMORY_BUFFER_DESC buf_desc = {
-            .offset = mem_info.offset,
-            .size = mem_info.size,
-            .flags = 0,
-        };
-        ret = CHECK_CU(cu->cuExternalMemoryGetMappedBuffer(&ebuf->buf, ebuf->mem, &buf_desc));
-        if (ret < 0)
-            goto error;
+    CUDA_EXTERNAL_SEMAPHORE_HANDLE_DESC s_desc = {
+#if HAVE_WIN32_DESKTOP
+        .type = IsWindows8OrGreater()
+            ? CU_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32
+            : CU_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_KMT,
+        .handle.win32.handle = evk->sync->signal_handle.handle,
+#else
+        .type = CU_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_FD,
+        .handle.fd = signal_fd,
+#endif
+    };
 
-        ra_vk_buf_set_user_data(buf, ebuf);
-    }
-    return buf;
+    ret = CHECK_CU(cu->cuImportExternalSemaphore(&evk->ss, &s_desc));
+    if (ret < 0)
+        goto error;
+    signal_fd = -1;
+
+    return true;
 
 error:
-    MP_ERR(mapper, "cuda_buf_pool_get failed\n");
-    return NULL;
+    MP_ERR(mapper, "cuda_ext_vk_init failed\n");
+    if (mem_fd > -1)
+        close(mem_fd);
+    if (wait_fd > -1)
+        close(wait_fd);
+    if (signal_fd > -1)
+        close(signal_fd);
+    return false;
 }
 
-static void cuda_buf_pool_uninit(struct ra_hwdec_mapper *mapper, int n)
+static void cuda_ext_vk_uninit(struct ra_hwdec_mapper *mapper, int n)
 {
     struct priv_owner *p_owner = mapper->owner->priv;
     struct priv *p = mapper->priv;
     CudaFunctions *cu = p_owner->cu;
 
-    struct ra_buf_pool *pool = &p->buf_pool[n];
-    for (int i = 0; i < pool->num_buffers; i++) {
-        struct ra_buf *buf = pool->buffers[i];
-        struct ext_buf *ebuf = ra_vk_buf_get_user_data(buf);
-        if (ebuf) {
-            if (ebuf->mem > 0) {
-                CHECK_CU(cu->cuDestroyExternalMemory(ebuf->mem));
-#if HAVE_WIN32_DESKTOP
-            }
-            if (ebuf->handle) {
-                // Handle must always be closed by us.
-                CloseHandle(ebuf->handle);
-            }
-#else
-            } else if (ebuf->fd > -1) {
-                // fd should only be closed if external memory was not imported
-                close(ebuf->fd);
-            }
-#endif
+    struct ext_vk *evk = &p->evk[n];
+    if (evk) {
+        if (evk->mem > 0) {
+            CHECK_CU(cu->cuDestroyExternalMemory(evk->mem));
+        }
+        if (evk->buf > 0) {
+            CHECK_CU(cu->cuMemFree(evk->buf));
+        }
+        if (evk->plbuf) {
```

Comment:
These non-NULL checks are technically unneeded, `pl_*_destroy` is safe to call on NULL.